File: skyspark-scala/src/main/scala/com/github/dkanellis/skyspark/scala/api/algorithms/bnl/Divider.scala
package com.github.dkanellis.skyspark.scala.api.algorithms.bnl

import com.github.dkanellis.skyspark.scala.api.algorithms.Point
import org.apache.spark.rdd.RDD

/**
  * Divides the data into appropriate partitions by using their flags and computes the local skylines for each
  * partition using the skyline computer.
  *
  * @param flagAdder              Adds the appropriate flags to the points
  * @param localSkylineCalculator Computes the local skylines
  */
private[bnl] class Divider(private val flagAdder: FlagAdder, private val localSkylineCalculator: LocalSkylineCalculator)
  extends Serializable {

  private[bnl] def this(bnlAlgorithm: BnlAlgorithm) = this(new FlagAdder, new LocalSkylineCalculator(bnlAlgorithm))

  private[bnl] def divide(points: RDD[Point]): RDD[(Flag, Point)] = {
    val flagPoints = flagAdder.addFlags(points)

    localSkylineCalculator.computeLocalSkylines(flagPoints)
  }
}

File: skyspark-java/src/main/java/com/akikanellis/skyspark/api/algorithms/bnl/Divider.java
package com.akikanellis.skyspark.api.algorithms.bnl;

import com.akikanellis.skyspark.api.algorithms.Point;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;

import java.io.Serializable;

class Divider implements Serializable {
    private final FlagAdder flagAdder;
    private final LocalSkylineCalculator localSkylineCalculator;

    Divider(FlagAdder flagAdder, LocalSkylineCalculator localSkylineCalculator) {
        this.flagAdder = flagAdder;
        this.localSkylineCalculator = localSkylineCalculator;
    }

    JavaPairRDD<Flag, Point> divide(JavaRDD<Point> points) {
        JavaPairRDD<Flag, Point> flagPoints = flagAdder.addFlags(points);

        return localSkylineCalculator.computeLocalSkylines(flagPoints);
    }
}
